MVVM Migration DONE

shriram-joshi opened the pull request.

shriram-joshi committed:
1) Started migration to MVVM.
2) Implemented mvvm for the customerShopsListFragment for the Recyclerview.
3) Have to work out the same for getting the pin code of the user.

shriram-joshi committed:
1) Implementation of mvvm for the customerShopsListFragment COMPLETE.
2) Changed the model of Customer to match the db.
3) This change includes creating 4 model pojo classes Customer.java, Address.java, Name.java and Credentials.java.
4) These classes were made from http://www.jsonschema2pojo.org/ using the JsonObject of a Customer from Firestore Database.
5) Ideally we should also convert the Vendor class to fit the database model.
6) I have created an issue for the same.

shriram-joshi committed:
1) Migrated customerSignUpDetails' Firestore part to mvvm.
2) Added a progressbar to fragment_customer_sign_up_details.xml

shriram-joshi committed:
1) Migration of LoginFragment to MVVM COMPLETE.
2) Created new Interface named OnAuthenticationListener for Authentication related functions in AuthenticationRepository.

shriram-joshi committed:
1) Migration of customerSignUpMain and customerSignUpDetails to mvvm COMPLETE.

shriram-joshi committed:
1) Migration of customerSignUpMain and customerSignUpDetails to mvvm COMPLETE.

shriram-joshi committed:
1) Migration of entire project to MVVM code architecture COMPLETE.
2) Also made the model POJOs compatible with the JsonObject returned by the database.
3) Added to progress bars wherever needed.

unc0ded commented:
Is there a typo perhaps in one of the commits? I see two commits with the same message, perhaps one of them was supposed to say vendor?

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/models/Address.java:
```diff
@@ -0,0 +1,74 @@
+    @SerializedName("city")
+    @Expose
+    private String city;
+    @SerializedName("addressLine1")
+    @Expose
+    private String addressLine1;
+    @SerializedName("addressLine2")
+    @Expose
+    private String addressLine2;
```
```suggestion
    @SerializedName("line1")
    @Expose
    private String line1;
    @SerializedName("line2")
    @Expose
    private String line2;
```

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/models/Address.java:
```diff
@@ -0,0 +1,74 @@
+        this.city = city;
+    }
+
+    public String getAddressLine1() {
+        return addressLine1;
+    }
+
+    public void setAddressLine1(String addressLine1) {
+        this.addressLine1 = addressLine1;
+    }
+
+    public String getAddressLine2() {
+        return addressLine2;
+    }
+
+    public void setAddressLine2(String addressLine2) {
+        this.addressLine2 = addressLine2;
+    }
```
```suggestion
    public String getLine1() {
        return line1;
    }

    public void setLine1(String line1) {
        this.line1 = line1;
    }

    public String getLine2() {
        return line2;
    }

    public void setLine2(String line2) {
        this.line2 = line2;
    }
```

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/repositories/VendorRepository.java:
```diff
@@ -0,0 +1,82 @@
+        vendorFdb.document(newVendor.getAddress().getPinCode()).get().addOnSuccessListener(documentSnapshot -> {
+            if (!documentSnapshot.exists()){
+                HashMap<String, Date> timeStamp = new HashMap<>();
+                timeStamp.put("New pin code available:", Calendar.getInstance().getTime());
```
Just as good practice, I think we shouldn't use the legacy Calendar class of java. It actually has many loopholes. Try this: 
```suggestion
                timeStamp.put("New pin code available:", LocalDateTime().now());
```

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/views/activities/SplashScreen.java:
```diff
@@ -20,6 +20,6 @@ public void run() {
                 finish();
             }
-        }, 500);
+        }, 0); //TODO: Change this back to 500ms after testing
```
It is irritating, isn't it? 😂

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/views/fragments/LoginFragment.java:
```diff
@@ -0,0 +1,276 @@
+            @Override
+            public void onVerificationFailed(@NonNull FirebaseException e) {
+                binding.progressBar.setVisibility(View.GONE);
+                Log.d("customerCallbacks", "onVerificationFailed", e);
+                Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
+                binding.emailOrPhone.setEnabled(true);
+                if (e instanceof FirebaseAuthInvalidCredentialsException) {
+                    Log.d("customerCallbacks","onVerificationFailed: Invalid Number");
+                }
+                else if (e instanceof FirebaseTooManyRequestsException) {
+                    Log.d("customerCallbacks","onVerificationFailed: SMS quota exceeded");
+                }
+            }
+
+            @Override
+            public void onCodeSent(@NonNull String s, @NonNull PhoneAuthProvider.ForceResendingToken forceResendingToken) {
+                binding.progressBar.setVisibility(View.GONE);
+                super.onCodeSent(s, forceResendingToken);
+                verificationId = s;
+
+                Toast.makeText(requireContext(), "OTP has been sent", Toast.LENGTH_SHORT).show();
+
+                OtpWidget otpView = new OtpWidget(requireContext());
+                otpView.getVerifyBtn().setOnClickListener(view -> {
+                    if (Objects.requireNonNull(otpView.getOtpView().getText()).toString().length() == 6) {
+                        PhoneAuthCredential customerCredential = PhoneAuthProvider.getCredential(verificationId, otpView.getOtpView().getText().toString());
+                        loginActivityVM.signInWithPhone(customerCredential);
+                        enter_otp.hide();
+                    }
+                    else
+                        Toast.makeText(requireContext(), "Please enter complete OTP", Toast.LENGTH_SHORT).show();
+                });
+                enter_otp = new MaterialAlertDialogBuilder(requireContext())
+                        .setTitle("Enter OTP")
+                        .setCancelable(true)
+                        .setView(otpView)
+                        .show();
+            }
+        };
+    }
```
Can't this whole part go into the Repository too? As for the UI tasks, those can be done using LiveData flags/listeners I guess.

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/views/fragments/customerSignUpMain.java:
```diff
@@ -115,20 +162,14 @@ else if (e instanceof FirebaseTooManyRequestsException) {
+                binding.otp.setEnabled(true);
+                verificationId = s;
             }
         };
```
Same question as before, about the Repository part.

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/views/fragments/customerSignUpMain.java:
```diff
@@ -91,19 +128,29 @@ public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstan
         customerCallbacks = new PhoneAuthProvider.OnVerificationStateChangedCallbacks() {
             @Override
             public void onVerificationCompleted(@NonNull PhoneAuthCredential phoneAuthCredential) {
+                binding.progressbar.setVisibility(View.GONE);
                 Log.d("customerCallbacks", "onVerificationCompleted:" + phoneAuthCredential);
-                signInWithPhoneAuthCredentials(phoneAuthCredential);
+                loginActivityVM.signUpWithPhone(phoneAuthCredential);
             }
 
             @Override
             public void onVerificationFailed(@NonNull FirebaseException e) {
+                binding.progressbar.setVisibility(View.GONE);
                 Log.d("customerCallbacks", "onVerificationFailed", e);
                 Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
-                binding.phoneNumber.setEnabled(true);
-                binding.otpBtn.setEnabled(true);
                 if (e instanceof FirebaseAuthInvalidCredentialsException) {
                     Log.d("customerCallbacks","onVerificationFailed: Invalid Number");
                 }
```
Same question as before, about the Repository part.

unc0ded commented on app/src/main/java/com/unc0ded/shopdeliver/views/fragments/vendorInventoryFragment.java:
```diff
@@ -71,27 +70,30 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
+                    .addSnapshotListener(((value, error) -> {
+                        if (error != null)
+                            Log.e("Failed", error.toString());
+                        else {
+                            int size = inventoryArray.size();
+                            for (int i = 0; i < size; i++) {
+                                inventoryArray.remove(0);
+                            }
+                            for (QueryDocumentSnapshot documentSnapshot: value) {
+                                inventoryArray.add(new Gson().toJsonTree(documentSnapshot.getData()).getAsJsonObject());
+                            }
+                            adapter.notifyDataSetChanged();
                         }
-                        for (QueryDocumentSnapshot documentSnapshot: value) {
-                            inventoryArray.add(new Gson().toJsonTree(documentSnapshot.getData()).getAsJsonObject());
-                        }
-                        adapter.notifyDataSetChanged();
-                    }
-                }));
+                    }));
+        }
     }
```
Is this part of the migration left? Because I did see an empty VendorMainActvityViewModel, and no repo for Vendor.

shriram-joshi replied:
Yep I'll reduce that np

unc0ded commented on app/src/main/res/layout/otp_dialog.xml:
```diff
@@ -17,7 +17,7 @@
         android:textColor="@color/line_colors"
         app:itemCount="6"
-        app:itemRadius="5dp"
+        app:itemRadius="20dp"
```
Doesn't this look better with a lesser corner radius?

shriram-joshi replied:
> I didn't see a Repository for the Vendor, where the fetching of inventory etc would be done? Has that intentionally not been done at this point?

I have left it out on purpose becuase I wasnt sure if the add to inventory way was final. I'll add it if its final.

unc0ded reviewed (changes requested):
I didn't see a Repository for the Vendor, where the fetching of inventory etc would be done? Has that intentionally not been done at this point?

shriram-joshi commented:
> Is there a typo perhaps in one of the commits? I see two commits with the same message, perhaps one of them was supposed to say vendor?

No I mistakenly forgot to make some changes before commiting a few things. They were some minor changes so I just made a new commit with the same message. Sorry for the confusion.

unc0ded commented:
Okay cool, just make a re-request after you commit the latest changes.

shriram-joshi commented:
> Okay cool, just make a re-request after you commit the latest changes.

I'll do it by tonight latest

shriram-joshi committed:
1) Migrated the inventory mechanism to MVVM.
2) Migration of entire project to MVVM code architecture COMPLETE.
3) Also made the requested changes from the code review.

shriram-joshi commented:
Making a new pull request for same purpose as this wasnt a draft and didnt include the latest commits I made today
